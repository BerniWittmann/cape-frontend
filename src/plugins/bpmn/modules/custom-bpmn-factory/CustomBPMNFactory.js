import inherits from 'inherits'

import BpmnFactory from 'bpmn-js/lib/features/modeling/BpmnFactory'

import { isCustomElement } from '../utils'

/**
 * A custom factory that knows how to create BPMN _and_ custom elements.
 */
export default function CustomBPMNFactory(moddle) {
  this._model = moddle
  BpmnFactory.call(this, moddle)
}

inherits(CustomBPMNFactory, BpmnFactory)

CustomBPMNFactory.$inject = [
  'moddle'
]

CustomBPMNFactory.prototype._needsId = function (element) {
  if (isCustomElement(element)) return true

  return BpmnFactory.prototype._needsId(element)
}
