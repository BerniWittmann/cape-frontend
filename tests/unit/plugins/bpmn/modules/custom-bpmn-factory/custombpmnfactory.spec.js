import CustomBPMNFactory from '@/plugins/bpmn/modules/custom-bpmn-factory/CustomBPMNFactory'
import BPMNFactory from 'bpmn-js/lib/features/modeling/BpmnFactory'

jest.mock('inherits', () => {
  return jest.fn()
})

jest.mock('bpmn-js/lib/features/modeling/BpmnFactory', () => {
  return {
    call: jest.fn(),
    prototype: {
      _needsId: jest.fn()
    }
  }
})

describe('Plugins', () => {
  describe('BPMN', () => {
    describe('Modules', () => {
      describe('Custom BPMN Factory', () => {
        const moddle = jest.fn()
        it('can be instantiated', () => {
          expect(CustomBPMNFactory).toEqual(expect.any(Function))
          CustomBPMNFactory.call({}, moddle)
          expect(BPMNFactory.call).toHaveBeenCalledWith({
            _model: moddle
          }, moddle)
        })

        it('has the correct modules injected', () => {
          expect(CustomBPMNFactory.$inject).toMatchSnapshot()
        })

        describe('checks wether an element needs an id', () => {
          const element = {
            $type: 'cape:ExtensionArea'
          }
          it('returns true when its a custom element', () => {
            expect(CustomBPMNFactory.prototype._needsId(element)).toBeTruthy()
            expect(BPMNFactory.prototype._needsId).not.toHaveBeenCalled()
          })
          describe('it relies the check logic back to BPMNFactory for non custom elements', () => {
            beforeEach(() => {
              element.$type = 'bpmn:Task'
            })
            it('can return true', () => {
              BPMNFactory.prototype._needsId = jest.fn().mockReturnValue(true)

              expect(CustomBPMNFactory.prototype._needsId(element)).toBeTruthy()
              expect(BPMNFactory.prototype._needsId).toHaveBeenCalled()
            })
            it('can return false', () => {
              BPMNFactory.prototype._needsId = jest.fn().mockReturnValue(false)

              expect(CustomBPMNFactory.prototype._needsId(element)).toBeFalsy()
              expect(BPMNFactory.prototype._needsId).toHaveBeenCalled()
            })
          })
        })
      })
    })
  })
})
