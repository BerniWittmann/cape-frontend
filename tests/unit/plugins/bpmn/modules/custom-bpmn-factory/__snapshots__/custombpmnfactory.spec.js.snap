// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Plugins BPMN Modules Custom BPMN Factory has the correct modules injected 1`] = `
Array [
  "moddle",
]
`;
